docs/proposals/metadata.md

```diff
@@ -0,0 +1,80 @@
+
+## Abstract
+
+A proposal for the additional metadata for Docker images which will provide more
+context about the content of the image and define relations between Docker
+images.
+
+## Motivation
+
+Image metadata are needed to do automatic generation of the OpenShift v3
+resource templates, like PodTemplates, ReplicationControllers, BuildConfigs
+etc.  Automatic generation of a PodTemplate is also required for
+Deployments.
+
+## Constraints and Assumptions
+
+For the time being, there is no good way to define metadata for Docker
+images other than specifying extra environment variables. There is upstream
+work currently being done to fix this situation and we should switch to that in
+the implementation in the future.
+
+This document only defines the metadata needed by the current set of use cases.
+Additional metadata and/or use cases may be added in the future.
+
+## Use cases
+
+1. As an author of Docker image that is going to be consumed by the OpenShift
+   platform, I want to express what categories/tags my Docker image will belongs
+   to, so the platform then can then consume the tags to improve the generation
+   workflow based on them.
+
+2. As an user of OpenShift I want to get reliable suggestions about services
+   that the Docker image I'm going to use might require to operate properly. As
+   an author of the Docker image, I need to have a way to record what services my
+   Docker image want to consume.
+
+2. As an author of the Docker image, I need to have a way to indicate whether
+   the container started from my Docker image does not support scaling.
+   The UI should then communicate this information with the end consumers.
+
+## Metadata variables (required for 3.0)
+
+### TAGS
+
+This metadata variable contains a list of tags which maps to the ports that the
+Docker image expose (EXPOSE instruction). The TAGS provides the context to the
+exposed ports and helps UI and generation tools to suggest relevant Docker
+images during the application creation process.
+
+Example format:
+
+```
+TAGS=database,mysql
+```
+
+### WANTS
+
+Specifies a list of tags that the generation tools and UI might use for
+suggesting to provide if you don't have the Docker images with given tags
+already. For example, if the Docker image WANTS 'mysql' and 'redis' and you
+don't have the Docker image with 'redis' tag, then UI might suggest you to add
+this image into your application.
+
+The format of the values is the same as for TAGS.
+
+## Metadata variables (post 3.0)
+
+### SINGLETON
+
+An image might use this variable to suggest that it does not support scaling.
+The UI will then communicate this to consumers of that image.
+Being not-scalable basically means that the value of 'replicas' should initially
+not be set higher than 1.
+
+### MIN_CPU,MIN_MEMORY,MIN_RESOURCE
+
+This variable suggests how much resources the Docker image might need in order
+to work properly. The UI might warn the user that deploying this Docker image
```

Comment:
I don't want the linkage aspect of publishes, I want the port metadata.  It's additions to EXPOSE.  I would call it out as something that is very specifically as additional expose metadata.

----- Original Message -----
> > +
> > +The format of the values is the same as for TAGS.
> > +
> > +## Metadata variables (post 3.0)
> > +
> > +### SINGLETON
> > +
> > +An image might use this variable to suggest that it does not support
> > scaling.
> > +The UI will then communicate this to consumers of that image.
> > +Being not-scalable basically means that the value of 'replicas' should
> > initially
> > +not be set higher than 1.
> > +
> > +### MIN_CPU,MIN_MEMORY,MIN_RESOURCE
> > +
> > +This variable suggests how much resources the Docker image might need in
> > order
> > +to work properly. The UI might warn the user that deploying this Docker
> > image
> 
> I thought you want to nuke PUBLISHES?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/906/files#r24571631